### ISSUE TITLE
New feature: Zun (Containers Service)

### ISSUE THREAD

**Issue by kevin-zhaoshuai** [gophercloud/gophercloud#700]
Current is [Zun](https://wiki.openstack.org/wiki/Zun) a Containers Service in OpenStack.  All supported stable Rest API is [here](https://developer.openstack.org/api-ref/application-container/). Gophercloud doesn't support Zun now.

Zun has two API entrypoint. One is "v1", the other is "experimental". 
The v1 is stable API for container management, also "experimental" API is for Container Capsule, which is mostly like "Pod" in Kubernetes, or "Container Group" in Azure container instance. Link of all info for [experimental API](https://review.openstack.org/#/c/484602/).

Now for zun, we want to join the project "[virtual-kubelet](https://github.com/virtual-kubelet/virtual-kubelet)", connecting OpenStack with Kubernetes. The project needs Zun to have a supported Go client. So now we plan to implement the "experimental" API first, then implement the v1  API.

For experimental API, we want to implement:
capsule/create
capsule/delete
capsule/show
capsule/describe

For v1 API, there are 3 separated part. [Refs](https://developer.openstack.org/api-ref/application-container/)
Manage containers
Manage service
Manage host

ref:
(1) https://docs.openstack.org/zun/latest/contributor/capsule.html
(2) https://github.com/virtual-kubelet/virtual-kubelet/issues/22
(3) https://review.openstack.org/#/c/437759/19/specs/container-composition.rst

**Comment by jtopjian** [comment]
I think adding Zun is certainly valid. @jrperritt ?

One thing to keep in mind is that in-house knowledge of Zun is most likely very limited (personally, I only know Zun exists and nothing more). Therefore, it'll be very important that PRs for Zun strictly follow our style and contributing guide. This will help make code reviews easier. If we have to hunt for the required information ourselves, it could delay the PR by possibly several weeks until we find the time to do so.

**Comment by jrperritt** [comment]
I think it's valid insofar as it looks like an OpenStack service (I've never heard of it), but experimental APIs are probably a non-starter for merging into `master`. 

@kevin-zhaoshuai I can make you a `zun-exp` branch if you like, where you can commit the experimental API until it's stable. We did that a while back for `magnum`, which I think in hindsight was a good decision.

**Comment by kevin-zhaoshuai** [comment]
@jtopjian @jrperritt Thanks a lot!
I think zun-exp is a good solution for raft development, if experimental API is stable, I will move it to stable.
For making code reviewer easier, I will strictly follow the code style and contributing guide. Actually the API is most likely compute api for "servers", the difference is the management object is container instead of VM.

**Comment by jrperritt** [comment]
Great. It has been created: https://github.com/gophercloud/gophercloud/tree/zun-exp.

**Comment by kevin-zhaoshuai** [comment]
@jtopjian Zun has move capsule API to V1 now. So I wonder could I implement the capsule API in gophercloud from branch zun-exp to master?
Or just finish all the actions of capsule in zun-exp and then migrate to master ?

**Comment by jtopjian** [comment]
@kevin-zhaoshuai That's great news - congratulations.

I think either option is valid - whichever you'd like to do.

**Comment by kevin-zhaoshuai** [comment]
@jtopjian If so, I'd like to merge zun-exp to master first.
Do I need to apply for a PR to merge zun-exp to master, or you can help to merge it directly?
Thanks a lot !

**Comment by jtopjian** [comment]
Good question.

@jrperritt is that something you'd like to handle?

**Comment by jrperritt** [comment]
To follow up on this, we'll probably wait on merging it into `master` until we can run the acceptance tests.

**Comment by kevin-zhaoshuai** [comment]
@jtopjian Haha, in my opinion, the option 3 is the most easy and efficiency way to solve this problem. So if we choose 3, what should we do next?
Thanks

**Comment by jtopjian** [comment]
@kevin-zhaoshuai I also think option 3 is the easiest, but I trust @jrperritt's judgement on this. So if he would prefer 1 or 2, I'll do what I can to help make either one as easy and quick as possible to get things rolling.

**Comment by kevin-zhaoshuai** [comment]
@jtopjian OK Thanks~let's wait for @jrperritt 's opinion.

**Comment by jtopjian** [comment]
@kevin-zhaoshuai From reviewing the discussion in the other PRs, I think the main issue was resolving this:

https://github.com/gophercloud/gophercloud/pull/865

I've cherry picked that commit and opened #930. If you'd like, I can merge it.

However, I'm not sure if merging that will cause issues when we finally merge `zun-exp` with `master`. Let's wait for @jrperritt to add some input.

**Comment by jtopjian** [comment]
@jrperritt I gave Option 2 a good try in #930 but there's a good amount of commits required where it would be easier just to do the large merge in #929. 

IMO, Option 3 is the easiest way to go. Thoughts?

**Comment by hongbin** [comment]
I think Option 3 is an ideal path for us (zun team). Per my understanding, Zun was put to a separated branch because the "capsule" API is experimental at that time. For now, the "capsule" API is graduated and merged into the stable API. We committed to maintain the stability of this API and any backward-incompatible change requires a bump of API micro-version by zun's team policy. Therefore, I think it is reasonable to consider merging zun code into the master branch now.

**Comment by jtopjian** [comment]
It looks like it'll take a little work to get a clean merge to happen. I will look into this and open a PR.

**Comment by jtopjian** [comment]
I've merged #944 which gets Zun into the master branch. You folks can use this issue (#700) to house additional Zun PRs, but I would recommend making an itemized list for all API calls that you intend to implement as to give a status of progress.

An example of this can be seen here: https://github.com/gophercloud/gophercloud/issues/723

Alternatively you can open a new issue for each PR or group of API calls.

**Comment by hongbin** [comment]
Cool. Thanks for your help @jtopjian . Greatly appreciate it.

**Comment by kevin-zhaoshuai** [comment]
@jtopjian  Thanks for help! I will do that!

**Comment by hongbin** [comment]
@jtopjian zun is supported now. i guess we can close this one?

**Comment by jtopjian** [comment]
@hongbin The title of this issue is about supporting Zun, which could be interpreted as the [entire Zun API suite](https://developer.openstack.org/api-ref/application-container/) (similar to, say, #1429). 

So far only the Capsule part has been implemented. That's OK - we can change the title to reflect only Capsule. But I tried looking up the API reference for capsules and wasn't able to find it?

**Comment by hongbin** [comment]
@jtopjian right. There are two missing parts: the API reference for Capsule at Zun side, and the container implementation at Gophercloud side. We will work on them so agree with you. It makes sense to keep this issue open. Just disregard my previous comment.

**Comment by jtopjian** [comment]
@hongbin Sure, sounds good. 

I don't want to put pressure on implementing _all_ of Zun in Gophercloud - there's no rush at all. The benefit of keeping this issue open, regardless of how long it takes, is just to show others the status of things.

If you only wanted to finish off Capsule, I think "update" is the only remaining piece? Again, just as a way of showing the status of things, at least others will be able to see that there was a discussion about Capsule, but the rest of Zun is still open for others to work on.

**Comment by hongbin** [comment]
@jtopjian , understand. Yes, "update" is the only missing part for Capsule. This API is not implemented in Zun yet, and might be the hardest one to implement but it is definitely on the plan.

**Comment by jtopjian** [comment]
@hongbin ahh I see!

That means Capsule is fully implemented. So totally up to you if you want to leave this issue open for all of Zun or change it to just Zun Capsule and close it off.

**Comment by hongbin** [comment]
@jtopjian , I think you made a really good point that leaving this issue open can show the progress of work for Zun implementation. I would lean to leave this open now. Thanks for the advice.

**Comment by jamesba** [comment]
Hi, was looking at this. If I were to add support for the v1/containers API would that be something that would be of interest?
